### ISSUE TITLE
Unable to query system includes when Windows Null service is disabled

### ISSUE THREAD

**Issue by Lpern** [microsoft/vscode-cpptools#2863]
**Type: LanguageService**

**Describe the bug**
- OS and Version: Windows 10 Home
- VS Code Version: 1.29.1
- C/C++ Extension Version: 0.20.1
- Other extensions you installed (and if the issue persists after disabling them): NONE
- A clear and concise description of what the bug is.

Also, a good note is that the C:\MinGW\bin is also already in the computer's system path.

I recently added an SSD to my computer, and did a clean Windows Install. I then installed VSC (on said SSD) and when I tried to use the C++ VSC extension again, it has been unable to parse the include file, and refuses to highlight errors. I have tried manually adding the MinGW include path to the includePath, SystemIncludePath, ForceIncludePath, BrowsePath, I have been reading up on all the issue tickets on this site, and I have tried every solution I saw, but the intellisense engine fails to respond every time, and then defaults to the Tag Parser even if I have intellisenseFallback disabled.  

I will include some example logs with the logging level at "Information" and "Debug", and I will also include my user .json settings.

The odd thing is that this extension worked fine before I did a clean OS Install, and I didn't have to tweak any defaults. I was, however, using an older version of the extension at the time. If there were a way to revert to a previous version, I might do that.

**Information logs**

  Folder: C:/MINGW/INCLUDE/ will be indexed
IntelliSense Engine = Default.
Autocomplete is enabled.
Error squiggles are enabled.
Discovering files...
  Processing folder (recursive): C:/MINGW/INCLUDE/
  Discovering files: 417 file(s) processed
  0 file(s) removed from database
Done discovering files.
Populate include completion cache.
Checking for syntax errors: file:///c%3A/Storage/Code_Repos/test.cpp
Parsing remaining files...
  Parsing: 0 files(s) processed
Done parsing remaining files.
Failed to read response from server: 20
0 include path suggestion(s) discovered.
IntelliSense engine is not responding. Using the Tag Parser instead.
0 include path suggestion(s) discovered.
IntelliSense engine is not responding. Using the Tag Parser instead.
IntelliSense engine is not responding. Using the Tag Parser instead.
IntelliSense engine is not responding. Using the Tag Parser instead.
0 include path suggestion(s) discovered.
IntelliSense engine is not responding. Using the Tag Parser instead.
Checking for syntax errors: file:///c%3A/Storage/Code_Repos/test.cpp
IntelliSense engine is not responding. Using the Tag Parser instead.
0 include path suggestion(s) discovered.
IntelliSense engine is not responding. Using the Tag Parser instead.
IntelliSense engine is not responding. Using the Tag Parser instead.
IntelliSense engine is not responding. Using the Tag Parser instead.
IntelliSense engine is not responding. Using the Tag Parser instead.
0 include path suggestion(s) discovered.

**Debug logs**

initialized
workspace/didChangeConfiguration
IntelliSense Engine = Default.
Autocomplete is enabled.
Error squiggles are enabled.
File exclude: **/.git
File exclude: **/.svn
File exclude: **/.hg
File exclude: **/CVS
File exclude: **/.DS_Store
File exclude: **/.vscode
Search exclude: **/node_modules
Search exclude: **/bower_components
Search exclude: **/.vscode
cpptools/queryCompilerDefaults
Attempting to get defaults from compiler found on the machine: ''
Attempting to get defaults from compiler found on the machine: C:\MinGW\bin\gcc.exe
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include/c++"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include/c++/mingw32"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include/c++/backward"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include"
ignoring duplicate directory "/mingw/lib/gcc/mingw32/6.3.0/../../../../include"
ignoring duplicate directory "c:/mingw/lib/gcc/../../include"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include-fixed"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/../../../../mingw32/include"
ignoring duplicate directory "/mingw/include"
#include "..." search starts here:
#include <...> search starts here:
 c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include/c++
 c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include/c++/mingw32
 c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include/c++/backward
 c:\mingw\bin\../li...
cpptools/didChangeFolderSettings
Attempting to get defaults from compiler found on the machine: C:\MinGW\bin\gcc.exe
Code browsing service initialized
Attempting to get defaults from compiler found on the machine: C:\MinGW\bin\gcc.exe
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include/c++"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include/c++/mingw32"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include/c++/backward"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include"
ignoring duplicate directory "/mingw/lib/gcc/mingw32/6.3.0/../../../../include"
ignoring duplicate directory "c:/mingw/lib/gcc/../../include"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include-fixed"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/../../../../mingw32/include"
ignoring duplicate directory "/mingw/include"
#include "..." search starts here:
#include <...> search starts here:
 c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include/c++
 c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include/c++/mingw32
 c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include/c++/backward
 c:\mingw\bin\../li...
  Folder: C:/STORAGE/CODE_REPOS/ will be indexed
Discovering files...
  Processing folder (recursive): C:/STORAGE/CODE_REPOS/
  Discovering files: 1 file(s) processed
  0 file(s) removed from database
Done discovering files.
Populate include completion cache.
Parsing remaining files...
  Parsing: 0 files(s) processed
Done parsing remaining files.
textDocument/codeAction
cpptools/activeDocumentChange
cpptools/textEditorSelectionChange
cpptools/textEditorSelectionChange
textDocument/didOpen
Unable to resolve include path: C:/MinGW/include
sending compilation args for C:\STORAGE\CODE_REPOS\TEST.CPP
  define: _DEBUG
  define: UNICODE
  define: _UNICODE
  stdver: --c++11
  intelliSenseMode: msvc
Checking for syntax errors: file:///c%3A/Storage/Code_Repos/test.cpp
queue_update_intellisense for files in tu of: C:\STORAGE\CODE_REPOS\TEST.CPP
textDocument/codeAction
Failed to read response from server: 20
textDocument/hover
IntelliSense engine is not responding. Using the Tag Parser instead.
using Tag Parser for quick info
textDocument/documentHighlight
IntelliSense client is currently disconnected
cpptools/textEditorSelectionChange
textDocument/codeAction
textDocument/hover
IntelliSense client is currently disconnected
IntelliSense engine is not responding. Using the Tag Parser instead.
using Tag Parser for quick info
textDocument/hover
IntelliSense client is currently disconnected
IntelliSense engine is not responding. Using the Tag Parser instead.
using Tag Parser for quick info
textDocument/hover
IntelliSense client is currently disconnected
IntelliSense engine is not responding. Using the Tag Parser instead.
using Tag Parser for quick info

**Test Code Used**

#include \<iostream>\
using namespace std;

int main(){
    cout << "Still not working" << endl;
    sdfas
}

**There was no error highlighting**

**User Settings (json)**

{
    
        //C++ extension configuration
    "C_Cpp.default.compilerPath": "C:\\MinGW/bin\\bin\\gcc.exe",
    //"C_Cpp.default.browse.path": [
        //"${workspaceRoot}\\**",
        //"C:\\MinGW\\include",
    //],
    "C_Cpp.default.cppStandard": "c++11",
    "C_Cpp.default.cStandard": "c11",
    //"C_Cpp.default.includePath": [
        //"C:\\MinGW\\include"
    //],
    "C_Cpp.errorSquiggles": "Enabled",
    "C_Cpp.intelliSenseEngine": "Default",
    "C_Cpp.workspaceSymbols": "All",
    //"C_Cpp.default.intelliSenseMode": "gcc-x64",
    "C_Cpp.intelliSenseEngineFallback": "Disabled",
    
    //"C_Cpp.default.forcedInclude": [
        //"C:\\MinGW\\include"
    //],
    //"C_Cpp.default.systemIncludePath": [
     //   "C:\\MinGW\\include"
    //],
    "C_Cpp.loggingLevel": "Debug",
        //end C++ entension configuration

    "terminal.integrated.shell.windows": "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe",
    "extensions.autoUpdate": false,
    "extensions.autoCheckUpdates": false,
}

As additional information, any time my cursor went to the include line, it decided to repeat the "switching to tag..." information output.

The commented out lines in the json, includes things I've tried in all combinations.

**Here is the Project's c_cpp_properties.json**

{
    "configurations": [
        {
            "name": "Win32",
            "compilerPath": "C:\\\\MinGW\\\\bin\\\\gcc.exe",
            "intelliSenseMode": "gcc-x64",
            "includePath": [
                "${workspaceFolder}/\*\*",
                "C:\\\MinGW\\\include"
            ],
            "defines": [],
            "cStandard": "c11",
            "cppStandard": "c++11",            
            "C_Cpp.intelliSenseEngineFallback": "Disabled"
        }
    ],
    "version": 4
}

**I also tried the above with forward slashes in the file paths**
**I could be doing something really stupid, or missing an obvious problem, so if anyone could give me a hand I'd very much appreciate it.**

**EDIT: So, I realized, that if I restarted it, that the program would show green squiggles around the include, and (in the error section) say "could not find includes in browse.path" but then as soon as my mouse cursor hovered over it, the squiggles went away, and the extension switched to tag parser... Which is weird.**

**Comment by Lpern** [comment]
**EDIT: Fixed Browse Path Issue:: IntelliSense still not working**

{
    
        //C++ extension configuration
    "C_Cpp.default.compilerPath": "C:\\MinGW\\bin\\gcc.exe",
    "C_Cpp.default.browse.path": [
        "${workspaceRoot}\\**",
        "C:\\MinGW\\include\\**",
        "C:\\MinGW\\lib\\gcc\\mingw32\\6.3.0\\include\\**"
    ],
    "C_Cpp.default.cppStandard": "c++11",
    "C_Cpp.default.cStandard": "c11",
    "C_Cpp.default.includePath": [
        "C:\\MinGW\\include",
        "C:\\MinGW\\lib\\gcc\\mingw32\\6.3.0\\include\\**"
    ],
    "C_Cpp.errorSquiggles": "Enabled",
    "C_Cpp.intelliSenseEngine": "Default",
    "C_Cpp.workspaceSymbols": "All",
    "C_Cpp.default.intelliSenseMode": "gcc-x64",
    "C_Cpp.intelliSenseEngineFallback": "Disabled",
    
    //"C_Cpp.default.forcedInclude": [
        //"C:\\MinGW\\include"
    //],
    //"C_Cpp.default.systemIncludePath": [
     //   "C:\\MinGW\\include"
    //],
    "C_Cpp.loggingLevel": "Information",
        //end C++ entension configuration

    "terminal.integrated.shell.windows": "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe",
    "extensions.autoUpdate": false,
    "extensions.autoCheckUpdates": false,
    "explorer.confirmDelete": false,
}

**Above is new User Settings (json)**

**New Debug Output**

initialized
workspace/didChangeConfiguration
IntelliSense Engine = Default.
Autocomplete is enabled.
Error squiggles are enabled.
File exclude: **/.git
File exclude: **/.svn
File exclude: **/.hg
File exclude: **/CVS
File exclude: **/.DS_Store
File exclude: **/.vscode
Search exclude: **/node_modules
Search exclude: **/bower_components
Search exclude: **/.vscode
textDocument/didOpen
cpptools/queryCompilerDefaults
Attempting to get defaults from compiler found on the machine: ''
Attempting to get defaults from compiler found on the machine: C:\MinGW\bin\gcc.exe
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include/c++"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include/c++/mingw32"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include/c++/backward"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include"
ignoring duplicate directory "/mingw/lib/gcc/mingw32/6.3.0/../../../../include"
ignoring duplicate directory "c:/mingw/lib/gcc/../../include"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include-fixed"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/../../../../mingw32/include"
ignoring duplicate directory "/mingw/include"
#include "..." search starts here:
#include <...> search starts here:
 c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include/c++
 c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include/c++/mingw32
 c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include/c++/backward
 c:\mingw\bin\../li...
cpptools/activeDocumentChange
cpptools/textEditorSelectionChange
textDocument/codeAction
cpptools/didChangeFolderSettings
Attempting to get defaults from compiler found on the machine: C:\MinGW\bin\gcc.exe
Code browsing service initialized
Attempting to get defaults from compiler in "compilerPath" property: 'C:\MinGW\bin\gcc.exe'
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include/c++"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include/c++/mingw32"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include/c++/backward"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include"
ignoring duplicate directory "/mingw/lib/gcc/mingw32/6.3.0/../../../../include"
ignoring duplicate directory "c:/mingw/lib/gcc/../../include"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include-fixed"
ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/../../../../mingw32/include"
ignoring duplicate directory "/mingw/include"
#include "..." search starts here:
#include <...> search starts here:
 c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include/c++
 c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include/c++/mingw32
 c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include/c++/backward
 c:\mingw\bin\../li...
Attempting to get defaults from compiler found on the machine: C:\MinGW\bin\gcc.exe
  Folder: C:/STORAGE/CODE_REPOS/ will be indexed
  Folder: C:/MINGW/INCLUDE/ will be indexed
  Folder: C:/MINGW/LIB/ will be indexed
  Folder: C:/MINGW/LIBEXEC/ will be indexed
Discovering files...
textDocument/didOpen
  Processing folder (recursive): C:/STORAGE/CODE_REPOS/
  Processing folder (recursive): C:/MINGW/INCLUDE/
  Processing folder (recursive): C:/MINGW/LIB/
  Processing folder (recursive): C:/MINGW/LIBEXEC/
  Discovering files: 1498 file(s) processed
  0 file(s) removed from database
Done discovering files.
Populate include completion cache.
Parsing remaining files...
  Parsing: 0 files(s) processed
Done parsing remaining files.
textDocument/hover
IntelliSense client not available, using Tag Parser for quick info.
using Tag Parser for quick info
$/cancelRequest
sending compilation args for C:\STORAGE\CODE_REPOS\TEST.CPP
  include: C:\Storage\Code_Repos
  define: _DEBUG
  define: UNICODE
  define: _UNICODE
  stdver: --c++11
  intelliSenseMode: msvc
Checking for syntax errors: file:///c%3A/Storage/Code_Repos/test.cpp
queue_update_intellisense for files in tu of: C:\STORAGE\CODE_REPOS\TEST.CPP
Failed to read response from server: 20
textDocument/hover
IntelliSense engine is not responding. Using the Tag Parser instead.
using Tag Parser for quick info
textDocument/hover
IntelliSense client is currently disconnected
IntelliSense engine is not responding. Using the Tag Parser instead.
using Tag Parser for quick info
textDocument/hover
IntelliSense client is currently disconnected
IntelliSense engine is not responding. Using the Tag Parser instead.
using Tag Parser for quick info


**Notes:**
I also tried switching VSC and Extension versions, here are what I've tried (in all possible combinations of both):

VSC: 1.29.1, 1.30.0 (insiders), 1.27.2
cpptools: 0.20.0, 0.19.0, & 0.17.7


I don't know why it required the includes in the browse path, and I'm 90% sure it will just read these as duplicates when the initial file parse happens (because it will go off of defaults for the MinGW libraries), but this seems to have taken care of the "not in browse.path" problem... However, it is still saying that it cannot read response from the IntelliSense server, and I'm still not sure why it's not working. As an extra oddity it says that the intelliSenseMode is still "msvc-x64" rather than my specified "gcc-x64" in the workspace settings and user settings... I'm not sure if that's what it's supposed to say, or... *shrugs*

I'm a CS student and I have turned to this extension to get a majority of my schoolwork done when using C++ since the school requires the use of the gcc compiler which they run on their (I'm assuming RHEL OS server).

Tag Parser should do fine until I can get some help with this issue, but the intelliSense and debugging were an amazing help for me earlier on. So, thank you in advance, and also I still think that this is a fantastic piece of equipment for any programmer.

**TLDR: It's not a browse.path issue: something is wrong with the intelliSense "server" possibly?**

**Comment by Lpern** [comment]
**EDIT: intelliSenseEngine keeps crashing, even while only using Tag Parser, and eventually extension host kills the extension entirely after the 5th intellisense shutdown... This is going to be an issue if I can't even use Tag Parser (which looks like it's working normally till shutdown)...**

I'm starting to think that it's something along the lines of[ this ticket: 2435](https://github.com/Microsoft/vscode-cpptools/issues/2453)?

**And now, I've gotten this exact same configuration to actively work on my laptop, a standard HDD Win10 1-partition laptop, with exactly the same everything as my desktop, and it's not having this issue. Down to the same user settings... AND it says "IntelliSenseMode: gcc" on initalization.**

**I don't understand this bug**

**Comment by Lpern** [comment]
After browsing some more tickets... I think this might be part of the problem:
https://github.com/Microsoft/vscode-cpptools/issues/2779#issuecomment-436464123

I'm having a sinking suspicion that even though I've set the default compiler path, it's still finding the cl.exe that my Visual Studio Community 2017 uses when I do Unreal Engine 4 development... Or... Maybe It's using Unreal Engine's clang compiler??? 

Either way, I think I've narrowed down the problem to the fact that it is still using the **"msvc" intelliSenseMode** That doesn't make things much better though, because I'm not uninstalling a bunch of other developer tools that I use frequently to see if that helps. 

I have the "C_Cpp.default.intelliSenseMode" set to "gcc-x64" in user settings, and I have the "intellisenseMode" set to that as well in the project settings json. I'm not really sure what to do from here.

I hope someone finds something to help, and/or I get support from the devs here. Please.

**Comment by bobbrow** [comment]
Can you zip up and share your code folder?  Email it to `c_cpp_support @ microsoft . com` That would probably be the easiest way for us to look at this.

**Comment by Lpern** [comment]
> Can you zip up and share your code folder? Email it to `c_cpp_support @ microsoft . com` That would probably be the easiest way for us to look at this.



Sure, by code folder, do you mean the code I was testing with, or the Visual Studio Code files? 
I know the extensions are enclosed in the users folder (seperate form the VSC directory) And logs and caches are available in appdata... I'll send all of the above, but if you meant something else let me konw.
You'll recieve an email from: rober188@************.edu

**Comment by sean-mcmanus** [comment]
I think he means your CODE_REPOS folder. It looks like the IntelliSense process is crashing with test.cpp.

**Comment by Lpern** [comment]
> I think he means your CODE_REPOS folder. It looks like the IntelliSense process is crashing with test.cpp.

It's not just crashing with test.cpp, **it's crashing with any .cpp/.c/.h files.** The debug output, and Code_Repos folder is using the same test code that's in the initial comment on this ticket, but I'll include that in the email.

**Said .cpp/.c/.h files also do not need to have included headers, and still crash Intellisense.**

**Comment by Lpern** [comment]
I have sent the email.
I had to remove the compiled a.exe, however, because google auto-assigns batch, and executable files, to be "viruses"

Still, it's just a test .cpp file. **It has this issue with ANY C++ or C file. Basically as soon as I hover over any include or code, it either crashes the intellisense server, or switches to Tag Parser**


If you look at the debug dumps I've included, in the initial comment, I have ruled out the specific files, or includes, and have tweaked settings accordingly. I tried files that were on the same partition, different partition, full read-write access, less read-write access, I tried running VSC as admin, I tried elevating program priority, and the problem is still that it always switches to the tag-parser because intelliSense won't respond, displays "intelliSenseMode: msvc" instead of gcc, and crashes every time I hover over code... Any code... In any file... unfortunately. 

I have also tried the last 5 versions of this extension sequentially (restarting computer between uninsatllation of previous Extension, clearing the json files, etc.) and the last versions of Visual Studio Code, and VSC Insiders. This "Code_Repos" folder, also works fine on my other VSC installation on my laptop.

There is something going wrong with the IntelliSenseEngine getting hung up on using msvc (maybe prioritizes existing 4 cl.exe modules??? This doesn't make sense though, because I changed the user settings to default to gcc, and I manually pathed the compiler to MinGW) mode, or being unable to parse header files. I don't really know at this point, but if there's a way you can fix it, I'd very much appreciate it.

Thank you for assisting me in this. I really appreciate the support.

**Comment by bobbrow** [comment]
I didn't see a `c_cpp_properties.json` in the files you sent.  In any case, I wasn't able to reproduce your problem.

### Here are the unexpected behaviors I found in your logs:
1. Under `Attempting to get defaults from compiler in "compilerPath" property: 'C:\MinGW\bin\gcc.exe'` the extension is logging the output from the compiler.  This is supposed to go to stdout/stderr and we shouldn't be logging this to the Output window unless the compiler returned an error code during the query.  There may be something wrong with your MinGW compiler.

**Can you open a cmd window and run the following commands?**  The "echo" command should print out 0.
```
cd /d c:\MinGW\bin
gcc.exe -Wp,-v -E -xc -dD -x c++ nul
echo %errorlevel%
```

2. Under `sending compilation args for C:\STORAGE\CODE_REPOS\TEST.CPP` the following path is not in ALL CAPS: `include: C:\Storage\Code_Repos`.  On Windows, all paths are normalized to upper case so that we have a reliable format to work with.  Do you have case-sensitivity turned on for your code folder?  This won't work in our extension unless you are using the Windows Subsystem for Linux.  (We have a feature request open to support this with a setting outside of the WSL context).

**Comment by Lpern** [comment]
With Win10 they added a file-specific feature to Windows 10 which allows you to specify certain library paths being case-sensitive in NTSF partition format... I forgot that I set up an L2-cache tiering system on the same partition for my external drive... And... apparently the root folder (none of the children) had the case-sensitive pathing on... I don't really know why, but I'm going to try a new MinGW instance in a few days, and I'll report the results...

I never would have thought of that. I don't know if this will fix it, but thank you for the help either way :)

**Comment by Lpern** [comment]
With Win10 they added a file-specific feature which allows you to specify certain library paths being case-sensitive in NTSF partition format... I forgot that I set up an L2-cache tiering system on the same SSD for my external drive... And... apparently the (storage) folder had the case-sensitive pathing on... I don't really know why (separate partition), but I'm going to try a new MinGW instance in a non-case-sensitive path in a few days, and I'll report the results...

I never would have thought of that. I don't know if this will fix it, but thank you for the help either way :)

**Comment by sean-mcmanus** [comment]
Yeah, we're tracking non-WSL case-sensitive file system support with https://github.com/Microsoft/vscode-cpptools/issues/1994.

**Comment by Lpern** [comment]
**UPDATE:**

So, I went in and manually set the case sensitivity to be disabled in both the MinGW Directories, and the Code_Repos directory... And still the same issue (even after restart), and I know for a fact that my MinGW installation is correct, because I tried out compiling both c++ and c test cases, and they all worked. Additionally, the test case you asked me to try seemed a bit odd, because "nul" isn't a file, and therefor isn't compilable. It's a very basic MinGW installation only the compiler for C++ and C are included, and it has always worked for me in the past (including working with VSC, like on my laptop: as we speek).

After checking several depth-one folders, the only one that had the casesensitivity enabled was the one I was using to help the caching program. 

I then also disabled case sensitivity for the entire C:\ drive, just to be safe. It took some prodding, but I'm 100% certain that the case sensitivity shouldn't be an issue at this juncture.

     C:\Windows\system32>cd C:\

     C:\>fsutil.exe file setCaseSensitiveInfo C:\Storage disable
     Case sensitive attribute on directory C:\Storage is disabled.

     C:\>fsutil.exe file setCaseSensitiveInfo C:\Storage\Code_Repos disable
     Case sensitive attribute on directory C:\Storage\Code_Repos is disabled.

     C:\>fsutil.exe file setCaseSensitiveInfo C:\Storage\Code_Repos\.vscode disable
     Case sensitive attribute on directory C:\Storage\Code_Repos\.vscode is disabled.

     C:\>fsutil.exe file queryCaseSensitiveInfo C:\Storage\Code_Repos
     Case sensitive attribute on directory C:\Storage\Code_Repos is disabled.

     C:\>fsutil.exe file queryCaseSensitiveInfo C:\
     Case sensitive attribute on directory C:\ is disabled.


I even tried the trick described in ticket #1994 after reading it. The psuedo-libraries `/mnt/c/include` read, but didn't seem to make a difference.


I think one of the possible problems here, is that it says it's setting the intellisense mode to `msvc` and not 'gcc' like it should. I read in a previous ticket, that if you had a 'cl.exe' or another compiler already "installed" then it searches for those. I have two main clang compilers on my system. One for VS Community 2017, and one for UnrealEngine4. I don't think these would cause a problem, but I could be wrong.

**Latest Relevant Debug Output:**

     Done discovering files.
     Populate include completion cache.
     Checking for syntax errors: file:///c%3A/Storage/Code_Repos/test.cpp
     Parsing remaining files...
     Parsing: 0 files(s) processed
     Done parsing remaining files.
     sending compilation args for C:\STORAGE\CODE_REPOS\TEST.CPP
     include: C:\MNT\C\INCLUDE
     include: C:\MINGW\LIB\GCC\MINGW32\6.3.0\INCLUDE\C++
     include: C:\MINGW\LIB\GCC\MINGW32\6.3.0\INCLUDE\C++\MINGW32
     include: C:\MINGW\INCLUDE
     include: C:\MINGW\LIB\GCC\MINGW32\6.3.0\INCLUDE
     define: _DEBUG
     define: UNICODE
     define: _UNICODE
     stdver: --c++11
     intelliSenseMode: msvc
     queue_update_intellisense for files in tu of: C:\STORAGE\CODE_REPOS\TEST.CPP
     Failed to read response from server: 20

The other odd thing here is that it says `Parsing remaining files.   &    Parsing:0 files processed` and the "checking for syntax errors" filepath is in lowercase.

The odd thing is that line `cc1plus.exe: fatal error: nul: No such file or directory` that keeps popping up when MinGW dumps the debug output into the Extension's output log. I don't think it's case sensitivity, but something is causing the attempted compilation of a non-existent file. I would think that it would be trying to compile the file from the given workspace root, but I don't know why it wouldn't send that path.

So, if there's a way where I can force the extension to use the gcc precepts instead of the msvc ones, I'd like to know. I set the "C_cpp.default.intelliSenseMode":"gcc-x64", and that didn't help.

**Comment by bobbrow** [comment]
Did you try to run the commands I listed above?  The extension is not trying to compile anything.  It's trying to query the compiler for the system include path and defines.  If you use that command without passing in `nul`, the compiler will hang and wait for input.  `nul` is an easy way to send an empty file and break out of the input loop.

What version of MinGW are you using? (i.e. where did you get it from)

The trick in #1994 doesn't work anymore.  You can only turn on case sensitive file support by selecting a WSL compiler. I edited the issue to reflect this.

**Comment by Lpern** [comment]
Ok... So... This is what happened when I tried to compile "nul"


     Microsoft Windows [Version 10.0.17763.134]
     (c) 2018 Microsoft Corporation. All rights reserved.

     C:\Users\logna_6jenblj>cd C:\MinGW

     C:\MinGW>gcc.exe -Wp,-v -E -xc -dD -x c++ nul
     ignoring nonexistent directory "c:\mingw\bin\../lib/gcc/mingw32/6.3.0/../../../../mingw32/include"
     ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include/c++"
     ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include/c++/mingw32"
     ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include/c++/backward"
     ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include"
     ignoring duplicate directory "/mingw/lib/gcc/mingw32/6.3.0/../../../../include"
     ignoring duplicate directory "c:/mingw/lib/gcc/../../include"
     ignoring duplicate directory "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/include-fixed"
     ignoring nonexistent directory 
     "c:/mingw/lib/gcc/../../lib/gcc/mingw32/6.3.0/../../../../mingw32/include"
     ignoring duplicate directory "/mingw/include"
     #include "..." search starts here:
     #include <...> search starts here:
      c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include/c++
      c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include/c++/mingw32
      c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include/c++/backward
      c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include
      c:\mingw\bin\../lib/gcc/mingw32/6.3.0/../../../../include
      c:\mingw\bin\../lib/gcc/mingw32/6.3.0/include-fixed
     End of search list.
     cc1plus.exe: fatal error: nul: No such file or directory
     compilation terminated.

     C:\MinGW>echo %errorLevel%
     1

     C:\MinGW>

but when I compile any other file, it works fine.
And I'm using the most recent windows build from the http://www.mingw.org/
I think it's version 6.3.0. (although technically there's not really a "true" version).

If you have a different idea for where I should get the MinGW libraries, then I'll be happy to try it.

**Comment by Lpern** [comment]
.... I got it working


So this is real dumb. I looked it up online, and realized that you were just passing the results to the NULL windows virtual device.

I knew it existed, but I thought that maybe 'nul' was a configured argument by the compiler or something.

When I realized that. I went and checked my registry, and sure enough, the SYSTEM reg keys, the null driver's start key was a 4 (hex)... It should have been a 1. 

I don't know why it was defaulted to 4, but now that the driver is running ( and I manually started it with `sc start Null`. 

If anyone reaches this ticket, depending on your windows version, if you get this problem, try commands here http://servicedefaults.com/10/null/

This worked.... 
It runs smoothly now. 
Thank you guys for all the help, I'm sorry it turned out to be something so stupid.

**Comment by bobbrow** [comment]
Thanks for the update.  I wouldn't have suspected this.  We should probably update the extension to guard against the Null service being disabled.

**Comment by bobbrow** [comment]
I suspect that we could add code to `sc query Null` and check if the service is running, but it doesn't seem that anyone else has hit this in the past few years, so we'll close this issue.  This could potentially be done in the TypeScript portion of our code and we'd accept a PR.
